### PR TITLE
[0019] doc: Add data structure doc which also captures latest script refactoring changes

### PR DESCRIPTION
**Pull request by xxuejie**
Ref https://github.com/nervosnetwork/ckb/pull/344

### CODE REVIEW

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+Category: Standards Track
+Status: Proposal
+Author: Haichao Zhu, Xuejie Xiao
+Organization: Nervos Foundation
+Created: 2019-03-26
+---
+
+# Data Structures of Nervos CKB
+
+This documents list all the basic structures one may need to know in order to develop on CKB, or mine CKB coins.
```
Comment (Mine77) [inline rfcs/0000-data-structures/0000-data-structures.md]: Shall we delete this sentence?
I think it's not suitable for an RFC document anymore.

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+```json
+{
+    "capacity": 5000000,
+    "data": "0x",
+    "lock": {
+      "args": [],
+      "binary_hash": "0xa58a960b28d6e283546e38740e80142da94f88e88d5114d8dc91312b8da4765a",
+      "version": 0
+    }
```
Comment (Mine77) [inline rfcs/0000-data-structures/0000-data-structures.md]: missing a comma here

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+```
+
+
+
+### Description
+
+| Name          | Type       | Description                                                  |
+| :------------ | :--------- | :----------------------------------------------------------- |
+| `version`     | uint8      | **The version of the script.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
+| `binary_hash` | H256(hash) | **The hash of ELF formatted binary that contains an RISC-V based script. For space consideration, the actual script is either attached to current transaction as a dep cell, or included in as an embed field in current the transaction. In either case, the hash of actual binary(which is cell data part for a dep cell, and the content of the whole field for an embed field) should match binary_hash included here. The actual binary is loaded into an CKB-VM instance when they are specified upon the transaction verification. |
```
Comment (Mine77) [inline rfcs/0000-data-structures/0000-data-structures.md]: missing ** at the end of the first sentence

Comment (Mine77) [reply rfcs/0000-data-structures/0000-data-structures.md]: > For space consideration, the actual script is either attached to current transaction as a dep cell, or included in as an embed field in current the transaction.

typo

-> For **space efficiency consideration**, the actual script is either attached to the current transaction as a dep cell, or included in as an embed field **in the current transaction**.

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+| `args`        | [Bytes]    | **An array of arguments as the script input.** The arguments here are imported into the CKB-VM instance as input arguments for the scripts. Note that for lock scripts, the corresponding CellInput would have another args field which is appended to the array here to form the real input arguments. |
+
+
+
+More information about Script can be [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/master/docs/how-to-write-contracts.md#script-model).
```
Comment (Mine77) [inline rfcs/0000-data-structures/0000-data-structures.md]: > More information about Script can be [here]

typo

-> More information about Script can be **found** [here]

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    ],
+    "outputs": [
+      {
+        "capacity": 5000000,
+        "data": "0x",
+        "lock": {
+          "args": [],
+          "binary_hash": "0xa58a960b28d6e283546e38740e80142da94f88e88d5114d8dc91312b8da4765a",
+          "version": 0
+        }
```
Comment (Mine77) [inline rfcs/0000-data-structures/0000-data-structures.md]: missing a comma here

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+#### Transaction
+
+| Name              | Type                           | Description                                                  |
+| ----------------- | ------------------------------ | ------------------------------------------------------------ |
+| `version`         | uint32                         | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
+| `hash`            | H256(Hash)                     | **The hash of the transaction.** This also serve as the identifier of the transaction. |
+| `deps`            | [`outpoint`]                   | **An array of `outpoint` that point to the cells that are dependencies of this transaction.** Only live cells can be listed here. The cells listed are read-only. |
+| `inputs`          | [{`previsou_output` , `args`}] | **An array of {`previous_output`, `args`}.** |
+| `embeds`          | [Bytes]                        | **An array of one-time used RISC-V binaries that can be referenced by `scripts` included in the transaction. ** |
```
Comment (Mine77) [inline rfcs/0000-data-structures/0000-data-structures.md]: > **An array of one-time used RISC-V binaries that can be referenced by `scripts` included in the transaction. ** 

* The bold gramma seems not functioning. Delete the space before the latter **

->  **An array of one-time used RISC-V binaries that can be referenced by `scripts` included in the transaction.**

**Comment by Mine77** [comment]
Also do you need to update the readme of the repo as well?
adding the RFC to the table

File: rfcs/0000-data-structures/0000-data-structures.md (hunk 1 above)
Comment (xxuejie) [reply rfcs/0000-data-structures/0000-data-structures.md]: I think it serves as a summary of the doc, what do you think?

**Comment by xxuejie** [comment]
> Also do you need to update the readme of the repo as well?
> adding the RFC to the table

Will do this once @doitian approves this and assigns an RFC number.

File: rfcs/0000-data-structures/0000-data-structures.md (hunk 1 above)
Comment (xxuejie) [reply rfcs/0000-data-structures/0000-data-structures.md]: I've slightly revised this a bit, let me know what you think.

Comment (Mine77) [reply rfcs/0000-data-structures/0000-data-structures.md]: lgtm

**Comment by doitian** [comment]
Assign RFC# 0019

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+| `inputs`          | [{`previsou_output` , `args`}] | **An array of {`previous_output`, `args`}.** |
+| `embeds`          | [Bytes]                        | **An array of one-time used RISC-V binaries that can be referenced by `scripts` included in the transaction.** |
+| `previous_output` | `outpoint`                     | **A cell outpoint that point to the cells used as inputs.** Input cells are in fact the output of previous transactions, hence they are noted as `previous_output` here. These cells are referred through  `outpoint`, which contains the transaction `hash` of the previous transaction, as well as this cell's `index` in its transaction's output list. |
+| `args`            | [Bytes]                        | **Additional input arguments provided by transaction creator to make the execution of corresponding lock script success**. One example here, is that signed signatures might be include here to make sure a signature verification lock script passes. See [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/develop/docs/how-to-write-contracts.md) for how to program this part. |
```
Comment (janx) [inline rfcs/0000-data-structures/0000-data-structures.md]:
```suggestion
| `args`            | [Bytes]                        | **Additional input arguments provided by transaction creator to make the execution of corresponding lock script success**. One example here, is that signatures might be include here to make sure a signature verification lock script passes.|
```

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+#### Header
+
+(`header` is a sub-structure of `block` and `UncleBlock`.)
+
+| Name           | Type                | Description                                                  |
+| -------------- | ------------------- | ------------------------------------------------------------ |
+| `cellbase_id`  | H256(hash)          | **The hash of the Cellbase transaction.** Cellbase transaction is just like the coinbase transaction in Bitcoin. It's the transaction added by the miner who mined this block, by which the miner receives block reward for successfully mined the block. |
+| `difficulty`   | Bytes               | **The difficulty of the PoW puzzle.**                        |
+| `hash`         | H256(hash)          | **The block hash.** Note that although this value is returned by the JSON-RPC API, it is in fact NOT a part of the header. This value is actually calculated from the `Header` data. |
```
Comment (janx) [inline rfcs/0000-data-structures/0000-data-structures.md]: `hash` is not a field of block/header.

We should either 1. rename the rfc to e.g. 'JSON-RPC result format', or 2. remove values returned by JSON-RPC if it's about data structure in protocol. A mix of the two is confusing IMO.

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+#### Transaction
+
+| Name              | Type                           | Description                                                  |
+| ----------------- | ------------------------------ | ------------------------------------------------------------ |
+| `version`         | uint32                         | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
+| `hash`            | H256(Hash)                     | **The hash of the transaction.** This also serve as the identifier of the transaction. |
+| `deps`            | [`outpoint`]                   | **An array of `outpoint` that point to the cells that are dependencies of this transaction.** Only live cells can be listed here. The cells listed are read-only. |
+| `inputs`          | [{`previsou_output` , `args`}] | **An array of {`previous_output`, `args`}.** |
+| `embeds`          | [Bytes]                        | **An array of one-time used RISC-V binaries that can be referenced by `scripts` included in the transaction.** |
```
Comment (janx) [inline rfcs/0000-data-structures/0000-data-structures.md]: `embeds` will be removed.

**Comment by xxuejie** [comment]
> Assign RFC# 0019

Updated.

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+}
+```
+
+
+
+### Description
+
+| Name          | Type       | Description                                                  |
+| :------------ | :--------- | :----------------------------------------------------------- |
+| `version`     | uint8      | **The version of the script.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
```
[the remark this answers is not in the document]
Comment (xxuejie) [reply rfcs/0000-data-structures/0000-data-structures.md]: I've revised this, how about how?

File: rfcs/0000-data-structures/0000-data-structures.md [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+| `capacity` | uint64     | **The size of the cell.** When a new cell is generated (via transaction), one of the verification rule is `capacity ≥ len(capacity)+len(data)+len(type)+len(lock)`. This value also represents the balance of CKB coin, just like the `balance` field in the Bitcoin's UTXO. (E.g. Alice owns 100 CKB coins means she can unlock a group of cells that has 100 amount of `capacity` in total.) |
+| `data`     | Bytes      | **Arbitrary data.** This part is for storing states or scripts.  In order to make this cell valid on-chain, the data filled in this field should comply with the logics and rules defined by `type` or `lock`. |
+| `type`     | `Script`   | **A Script that defines the type of the cell.** In a transaction, if an input cell and an output cell has the same `type` field, then the `data` part of these two cells is limited by the `type` script upon the transaction verification. (I.e. `type` is a script that limits how the `data` field of the new cells can be changed from the old cells.) `type` is required to has a data structure of `script`. **This field is optional.** |
```
[the remark this answers is not in the document]
Comment (Mine77) [reply rfcs/0000-data-structures/0000-data-structures.md]: Do you mean that this is not required by the CKB protocol, but something commonly adpoted by type script?
I.e. the type script of an output would find the inputs that has the same type script has with it, and use these inputs data as the old state of this type of cell.

[the remark this answers is not in the document]
Comment (Mine77) [reply rfcs/0000-data-structures/0000-data-structures.md]: I think the description here does not imply that input's type should be equal to output type.

[the remark this answers is not in the document]
Comment (janx) [reply rfcs/0000-data-structures/0000-data-structures.md]: > In a transaction, if an input cell and an output cell has the same `type` field, then the `data` part of these two cells is limited by the `type` script upon the transaction verification.

The `data` in a cell is constrained by the logic in `type` no matter whether there's input/output cell has the same `type`. The 'if' here is an unnecessary condition.

File: rfcs/0009-vm-syscalls/0009-vm-syscalls.md [hunk 12]
```diff
@@ -227,13 +228,36 @@ The arguments used here are:
...
+This syscall would first locate an input in current transaction via `source` and `index` value, it then serialize the extracted field into flatbuffer format, then use the same steps as documented in *Load Transaction* syscall to feed data into VM.
 
 Specifying an invalid source value here would immediately trigger a VM error, however specifying `output` as the source here would only result in `2` as return value, specifying `current` as source in a *type* script, which doesn't have input, would also result in `2` as return value. Specifying an invalid index value here, would result in `2` as return value, denoting item missing state. Specifying any invalid field will also trigger VM error immediately. Otherwise the syscall would return `0` denoting success state.
 
+NOTE there is one quirk when requesting `args` part in `deps`: since CFB doesn't allow using a vector as the root type, we have to wrap `args` in a `CellInput` table, and provide the `CellInput` table as the CFB root type instead.
+
+### Load Embed
```
Comment (janx) [inline rfcs/0009-vm-syscalls/0009-vm-syscalls.md]: Can be removed?

File: rfcs/0019-data-structures/0019-data-structures.md [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+
+
+## Script
+
+### Example
+
+```json
+{
+  "version": 0,
```
Comment (janx) [inline rfcs/0019-data-structures/0019-data-structures.md]: `version` will be removed from `Script` cc @xxuejie

Comment (xxuejie) [reply rfcs/0019-data-structures/0019-data-structures.md]: Yes, but what about we removed this once the PR at CKB is crafted?

Comment (janx) [reply rfcs/0019-data-structures/0019-data-structures.md]: That's ok. We can merge this PR after CKB PR. No hurry.

File: rfcs/0019-data-structures/0019-data-structures.md [hunk 14]
```diff
@@ -0,0 +1,255 @@
+---
+Number: "0000"
```
Comment (u2) [inline rfcs/0019-data-structures/0019-data-structures.md]: Number.

Comment (doitian) [reply rfcs/0019-data-structures/0019-data-structures.md]: Looks like we need a script to check such errors via CI.

**Comment by xxuejie** [comment]
> Please cp the change to remove cellbase here [nervosnetwork/ckb@528e95e#diff-9cec9a07c6273e16f3fbf77943c66196](https://github.com/nervosnetwork/ckb/commit/528e95e845bcd98a51c91a7a353e531e36fdf6ae#diff-9cec9a07c6273e16f3fbf77943c66196) to avoid another PR.

This is done now

File: README.md [hunk 15]
```diff
@@ -69,6 +69,7 @@ The maintainers of RFCs will review the proposal, ask if there's any objections,
 | [13](rfcs/0013-get-block-template) | [Block Template](rfcs/0013-get-block-template/0013-get-block-template.md) | Dingwei Zhang | Standards Track | Proposal |
 | [14](rfcs/0014-vm-cycle-limits) | [VM Cycle Limits](rfcs/0014-vm-cycle-limits/0014-vm-cycle-limits.md) | Xuejie Xiao | Standards Track | Proposal |
 | [17](rfcs/0017-tx-valid-since) | [Transaction valid since](rfcs/0017-tx-valid-since/0017-tx-valid-since.md) | Jinyang Jiang | Standards Track | Proposal
+| [19](rfcs/0019-data-structures) | [Data Structures](rfcs/0019-data-structures/0019-data-structures.md) | Haichao Zhu, Xuejie Xiao | Standards Track | Proposal
```
Comment (janx) [inline README.md]:
```suggestion
| [19](rfcs/0019-data-structures) | [Data Structures](rfcs/0019-data-structures/0019-data-structures.md) | Haichao Zhu, Xuejie Xiao | Informational | Draft
```

`Informational` is more appropriate as this is an RFC reflects design decisions formed somewhere else.

File: rfcs/0019-data-structures/0019-data-structures.md [hunk 16]
```diff
@@ -0,0 +1,248 @@
+---
+Number: "0019"
+Category: Standards Track
```
Comment (janx) [inline rfcs/0019-data-structures/0019-data-structures.md]:
```suggestion
Category: Informational
```

File: rfcs/0019-data-structures/0019-data-structures.md [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+Status: Proposal
```
Comment (janx) [inline rfcs/0019-data-structures/0019-data-structures.md]:
```suggestion
Status: Draft
```

**Pull request merged**